### ISSUE TITLE
WP-Env: XDebug requires configuration to support WSL2

### ISSUE THREAD

**Issue by Clorith** [WordPress/gutenberg#30703]
## Description
When developing on Windows, using WSL2 (a proper Linux kernel running under Windows, so you get native performance gains to web applications for example) is becoming more and more common.

the `@wordpress/env` package by default works great under WSL2, but the xdebug configuration requires special attention.

The current configuration added when the `--xdebug` flag is present adds the line

```
xdebug.discover_client_host=true
```

This works great when the same host system is used, or when there's direct access, unfortunately with WSL2 there's a system in between, so the remote debugging session never connects to your IDE, but instead connects to the WSL2 system, and ends up in a black hole there.

The benefit to `discover_client_host` is that is' zero-config if you wish to share a debug session with others on the same network as you, but I wonder if this is actually needed, based on the pretense that this package is meant for local development, not deployments?

The solution, when using WSL2, is to disable `discover_client_host` and instead define the client host directly;

```
xdebug.discover_client_host=false
xdebug.client_host=host.docker.internal
```

It would also be possible to produce a conditional config here, there are a few ways to check if WSL is in use, one could check the `uname -a` value (My current output of this is `Linux DESKTOP-L81G2QH 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`, which includes a `WSL2` suffix), and also a environmental variable of `WSL_DISTRO_NAME` is always set on these systems.

---

That's a lot of background history, and then I looked at the source it self, and it looks like there is already a conditional check, the problem is that WSL is returning a Linux system to `os.type()`, which needs to be a false-y value for the `host.docker.internal` setting to be used, so perhaps this needs expanded to check for WSL if a Linux system is detected? Checking for `os.release()` alongside it would reveal the WSL build identifier, so maybe [init-config.js:101](https://github1.com/WordPress/gutenberg/blob/3a7cb5930475cf744c9bf6e170d0df3aa723af71/packages/env/lib/init-config.js#L101) should be something along these lines:

```
const isLinux = os.type() === 'Linux' && ! os.release().indexOf( 'WSL' )
```

Ref: https://github.com/WordPress/gutenberg/blob/3a7cb5930475cf744c9bf6e170d0df3aa723af71/packages/env/lib/init-config.js#L101-L108

**Comment by gaambo** [comment]
A more extended approach would be to allow adding extra environment variables or including custom php.ini settings via volume-mounting. That could also allow for more complex situations where you need special php extensions to be set.
If I could add another .ini file that get's mapped into `/usr/local/etc/php/` I could overwrite the xdebug config. 

My workaround ATM is to run this after start: 
```
wp-env run development "RUN echo 'xdebug.client_host=host.docker.internal'  >> /usr/local/etc/php/php.ini"
/etc/init.d/apache2 restart
```

**Comment by jawinn** [comment]
Thanks for that @gaambo , that saved me from a huge headache in trying to figure out why VS Code + WSL2 + the wp-env docker was not connecting / stopping at any breakpoints. A "black hole" indeed. I tried every config and recommended launch.json setting I could find until I found this thread.

Could this be documented in the [Xdebug section of the README ](https://github.com/WordPress/gutenberg/blob/trunk/packages/env/README.md#xdebug-ide-support)until this is fixed? Otherwise the instructions were correct.

Also for some reason running that echo command gave me 
`/bin/sh: 1: cannot create /usr/local/etc/php/php.ini: Directory nonexistent`.
So I had to open the container in a terminal and run the 
`echo "xdebug.client_host=host.docker.internal" >> /usr/local/etc/php/php.ini`.

**Comment by danieliser** [comment]
Not sure what is going on but I think docs for this and these commands above are not correct any longer.

Running the following does generate the `directory nonexistent` issue for me, but if I simply use some escaped quotes for the inner command it runs through but since I first started troubleshooting this a week ago I've never gotten it to work.

` wp-env run cli "sh -c \"echo 'xdebug.client_host=host.docker.internal'  >> /usr/local/etc/php/php.ini\""`

Just realized today though that i'm consistently getting a development service not found:

```
wp-env run development ls /
ℹ Starting 'ls /' on the development container. 

no such service: development
✖ Command failed with exit code 1
Command failed with exit code 1
```

Running the same on `cli` shows that it runs the command on both wordpress & mysql boxes, to which I get a `permission denied ` error on writing to php.ini. I believe this is because that file doesn't exist on mysql container.

Further the docs for `wp-env run [container]` lists containers as `development`, `tests`, or `cli`.

I can run commands against the container names from the generated docker.compose file, BUT, running the following seems to run it against the wrong container.

```
wp-env run wordpress  "sh -c \"echo 'xdebug.client_host=host.docker.internal'  >> /usr/local/etc/php/php.ini\""
ℹ Starting 'sh -c "echo 'xdebug.client_host=host.docker.internal'  >> /usr/local/etc/php/php.ini"' on the wordpress container. 

[+] Running 1/0
 ⠿ Container d17d9b1fbb97f4a1ea11c2ba3d4d1279-mysql-1  Running                                                                                                                        0.0s
✔ Ran `sh -c "echo 'xdebug.client_host=host.docker.internal'  >> /usr/local/etc/php/php.ini"` in 'wordpress'. (in 1s 418ms)
```
Notice specifically I ran against `wordpress` but it was processed in the `mysql` container.

Am I wrong, crazy or is this a bug that is newer?

**Comment by gaambo** [comment]
I just tested the merged #50490 in trunk and it actually fixes this problem. I wasn't expecting this to work in WSL2 since it actually removes the isLinux check but it works. 🎉

Environment:
- WSL2 with docker service/daemon installed inside wsl
- Gutenberg repository with trunk at current moment (after #50490 as merged)
- VSCode with WSL integration

Steps for positive test:
- Run `npm env start -- --xdebug` (need to use the npm script to use the latest wp-env version from the package itself and the `-- --` format for passing flags to it)
- Copy the `.vscode/launch-example.json` to `.vscode/launch.json` 
- Set a breakpoint in `gutenberg.php`
- Start the debugger/launch it in VSCode
- Open the site and see the breakpoint being hit.

Also the merge request enables us to map `.htaccess` files which could also be used to set php.ini values.

**Comment by skorasaurus** [comment]
closing since https://github.com/WordPress/gutenberg/pull/50490 apparently fixes it :) reopen if you're still experiencing this.
